src/core/createConfig.js

```diff
@@ -48,18 +51,46 @@ const createConfig = config => {
       return keys;
     },
     /**
-     * Adds schema information to the existing configuration schema.
+     * Adds more validators to any existing validators.
      */
-    extendSchema: schemaAddition => {
-      assign(cfg.schema, schemaAddition);
-      return cfg.schema;
+    addValidators: validators => {
+      assign(cfg.validators, validators);
+      return cfg.validators;
     },
     /**
-     * Validates the configuration against the defined schema.
+     * Validates the configuration against the defined validators.
      */
     validate: () => {
-      // TODO: Validate existing configuration against defined schema.
-      return true;
+      const keys = Object.keys(cfg.validators);
+      const errors = keys.reduce((ac, key) => {
+        const validator = cfg.validators[key];
+        const currentValue = cfg.get(key);
+        if (
+          !validator.validate &&
+          currentValue == null &&
+          validator.defaultValue
+        ) {
+          cfg.set(key, validator.defaultValue);
+        } else if (validator.validate) {
+          const errorMessage = validator.validate(
```

Comment:
Does the `validate`  function need the default value? Doesn't seem needed in `required` for example, which should have the same signature.

Reply:
If there is a default defined for a whole object structure you might want to access it in the validate function. That is, if someone has defined several configurations but not specified one default value it could be picked up and set by getting it from the default value passed in to validate.

Reply:
If I read this right, if the validator has both `defaultValue` and `validate` defined, the `defaultValue` won't be set on the configuration automatically like it would be if `validate` were not defined. Is that intended? If so, why?

Reply:
Sorry, it seems I hijacked the thread unintentionally.

Reply:
After your change, it now seems like `validator.validate` will only get called if default value is **not** applied. Is that intentional?

Reply:
Yes, if the default value is applied then validate is not called. That is if a default value has been specified (within a validator) I would expect that value to be valid.

Reply:
That makes sense. Thanks for explaining.